File: c++/20.cpp
// 20. Valid Parentheses
// Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.
//
// An input string is valid if:
//
// Open brackets must be closed by the same type of brackets.
// Open brackets must be closed in the correct order.
// Note that an empty string is also considered valid.
//
// Example 1:
//
// Input: "()"
// Output: true
// Example 2:
//
// Input: "()[]{}"
// Output: true
// Example 3:
//
// Input: "(]"
// Output: false
// Example 4:
//
// Input: "([)]"
// Output: false
// Example 5:
//
// Input: "{[]}"
// Output: true

// 0 ms, faster than 100.00%
class Solution {
public:
  bool isValid(string s) {
    int len = s.size();
    if (len % 2 == 1) return false;
    vector<char> p;
    for (int i = 0; i < len; i++) {
      if (s[i] == '(' || s[i] == '[' || s[i] == '{') {
        p.push_back(s[i]);
      }  else if ((s[i] == ')' || s[i] == ']' || s[i] == '}') && p.size() == 0) {
        return false;
      } else if (s[i] == ')' && p.back() == '(') {
        p.pop_back();
      } else if (s[i] == ']' && p.back() == '[') {
        p.pop_back();
      } else if (s[i] == '}' && p.back() == '{') {
        p.pop_back();
      }
    } 
    if (p.size() > 0) return false;
    return true;
  }
};

File: go/20.go
// 20. Valid Parentheses

// Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

// An input string is valid if:

// Open brackets must be closed by the same type of brackets.
// Open brackets must be closed in the correct order.
// Every close bracket has a corresponding open bracket of the same type.
 

// Example 1:

// Input: s = "()"

// Output: true

// Example 2:

// Input: s = "()[]{}"

// Output: true

// Example 3:

// Input: s = "(]"

// Output: false

// Example 4:

// Input: s = "([])"

// Output: true

 

// Constraints:

// 1 <= s.length <= 104
// s consists of parentheses only '()[]{}'.

// Runtime 0 ms Beats 100.00%
// Memory 4.11 MB Beats 64.76%
func isValid(s string) bool {
    stack := []rune{}
    for _, char := range s {
        if char == '(' || char == '[' || char == '{' {
            stack = append(stack, char)
        } else {
            if len(stack) == 0 {
                return false
            }
            top := stack[len(stack) - 1]
            if (char == ')' && top != '(') || (char == ']' && top != '[') || (char == '}' && top != '{') {
                return false
            }
            stack = stack[:len(stack) - 1]
        }
    }
    return len(stack) == 0
}
